### PR TITLE
Updating the kim citation report to lammps conventions

### PR DESCRIPTION
**Pull request by yafshar**
**Summary**

This PR includes updating the kim citation report to lammps conventions. 

**Related Issue(s)**

This is reported in https://github.com/lammps/lammps/pull/2620#issuecomment-784271608

**Author(s)**

Yaser Afshar (UMN)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Backward compatible

**Implementation Notes**

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/KIM/kim_init.cpp [hunk 1]
```diff
@@ -514,19 +523,25 @@ void KimInit::write_log_cite(char *model_name)
...
-        &availableAsString, &fileString);
+      collections, i, &fileName, nullptr, nullptr,
+      &availableAsString, &fileString);
     if (err) continue;
 
-    if (0 == strncmp("kimcite", fileName, 7)) {
-      if ((lmp->citeme) && (availableAsString)) lmp->citeme->add(fileString);
-    }
+    if ((strncmp("kimcite", fileName, 7) == 0) and availableAsString)
+      cite_openkim_potential += fileString;
```
Comment (yafshar) [inline src/KIM/kim_init.cpp:541]: @akohlmey is the concatenation safe here from your view or should I change it to `cite_openkim_potential += std::string(fileString);` or `cite_openkim_potential += fmt::format("{}", fileString);`

**Comment by rbberger** [comment]
@yafshar as much as I would like to use `std::regex`, note that with GCC  < 4.9.0 it is an empty shell. It compiles (!!!) but does random things since it wasn't actually implemented until 4.9.0 (see [this](https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions) for more details)

**Comment by yafshar** [comment]
> @yafshar as much as I would like to use `std::regex`, note that with GCC < 4.9.0 it is an empty shell. It compiles (!!!) but does random things since it wasn't actually implemented until 4.9.0 (see [this](https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions) for more details)

@rbberger I agree, that is why I asked about the LAMMPS interface in `utils regex-module`. The use of regex in this PR can simply be replaced with LAMMPS regex-module if you put it in the interface.

File: src/KIM/kim_init.cpp (hunk 1 above)
Comment (akohlmey) [reply src/KIM/kim_init.cpp:541]: yes, that is fine. I am more concerned about the use of `strncmp()`. That is error prone and difficult to debug. Better would be using `utils::strmatch(fileName,"^kimcite")` which will return true/false

**Comment by akohlmey** [comment]
> > @yafshar as much as I would like to use `std::regex`, note that with GCC < 4.9.0 it is an empty shell. It compiles (!!!) but does random things since it wasn't actually implemented until 4.9.0 (see [this](https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions) for more details)
> 
> @rbberger I agree, that is why I asked about the LAMMPS interface in `utils regex-module`. The use of regex in this PR can simply be replaced with LAMMPS regex-module if you put it in the interface.

For matching strings to regular expressions, you can use `utils::strmatch()`. Please note, that the embedded regex library is very minimal and has limitations. It is not intended to expose the entire library, but only utility functions using it. It is rather a stopgap measure until we can use a proper system library that is guaranteed to be available with the same interface for all supported platforms.

**Comment by yafshar** [comment]
> > > @yafshar as much as I would like to use `std::regex`, note that with GCC < 4.9.0 it is an empty shell. It compiles (!!!) but does random things since it wasn't actually implemented until 4.9.0 (see [this](https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions) for more details)
> > 
> > 
> > @rbberger I agree, that is why I asked about the LAMMPS interface in `utils regex-module`. The use of regex in this PR can simply be replaced with LAMMPS regex-module if you put it in the interface.
> 
> For matching strings to regular expressions, you can use `utils::strmatch()`. Please note, that the embedded regex library is very minimal and has limitations. It is not intended to expose the entire library, but only utility functions using it. It is rather a stopgap measure until we can use a proper system library that is guaranteed to be available with the same interface for all supported platforms.

@akohlmey  here I need to extract a simple pattern `"[MOS]{2}_[0-9]{12}_[0-9]{3}"` from the model name and thus I need more than checking

**Comment by akohlmey** [comment]
> @akohlmey here I need to extract a simple pattern `"[MOS]{2}_[0-9]{12}_[0-9]{3}"` from the model name and thus I need more than checking. This is very simple and totally possible with utils regex-module

Let me have a closer look and think about an interface.

**Comment by akohlmey** [comment]
@yafshar since we are cleaning up legacy C-style string processing for simpler/safer/readable alternatives, here is another opportunity:
```
 if (!(strcmp(arg[0], "create") == 0) &&
      !(strcmp(arg[0], "destroy") == 0) &&
      !(strcmp(arg[0], "modify") == 0) &&
      !(strcmp(arg[0], "remove") == 0) &&
      !(strcmp(arg[0], "dump") == 0)) {
```
could become
```
const std::string subcmd(arg[0]);
 if ((subcmd != "create") &&  (subcmd != "destroy") && (subcmd != "modify")
     && (subcmd != "remove") && (subcmd != "dump")) {
```
There are similar uses of `strcmp()` elsewhere in the KIM package. Not all of them make sense to refactor, but particularly cases where the same word is compared to strings multiple times would become more readable.

File: src/KIM/kim_init.cpp (hunk 1 above)
Comment (akohlmey) [reply src/KIM/kim_init.cpp:541]: On second thought, there is a problem with this code. Not in how the string is constructed, but the fact that the string is going out of scope. The citeme class currently stores the address of the string pointer in a set to identify strings that have already been submitted. But that will break with this use. I will have to change that part of `CiteMe::add()` to use a hash instead. Then I can also make it use a `const std::string &` as argument and having temporaries added would not be a problem.

**Comment by yafshar** [comment]
> There are similar uses of `strcmp()` elsewhere in the KIM package. Not all of them make sense to refactor, but particularly cases where the same word is compared to strings multiple times would become more readable.

I am cleaning all the possible cases as suggested and will commit them after testing.

File: src/KIM/kim_init.cpp (hunk 1 above)
Comment (akohlmey) [reply src/KIM/kim_init.cpp:542]: PR #2629 should address my concerns about temporaries.

**Comment by akohlmey** [comment]
@yafshar I have updated the PR to use the new `utils::strfind()` function I just added. Mind you, the mini-regex code does not support `{#}`, so I had to adjust the regex string with explicit matches (but it has "\d" for digits). It seems to be working as expected for me and passes the available unit tests. Please check it out and make (final?) adjustments, if needed. Then this could be merged. Also the update to the CiteMe class is included, so it should work reliably with the generated citation strings.

**Comment by yafshar** [comment]
@akohlmey thanks for the update. I check the code and do all my testing.

**Comment by yafshar** [comment]
@akohlmey I did my tests and everything seems good to me. Thanks to your modification the citations are correct and the string is not going out of scope.

**Review by sjplimp — APPROVED** [review]

**Review by akohlmey — APPROVED** [review]

**Pull request merged**